ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForStructTest.java

```diff
@@ -0,0 +1,156 @@
+    assertThat(querySpecification.getSelect().getSelectItems().size(), equalTo(1));
+    final Expression col0 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(0))
+        .getExpression();
+
+    assertThat(col0, instanceOf(DereferenceExpression.class));
+    assertThat(col0.toString(), equalTo("ORDERS.ORDERID"));
+  }
+
+
+  @Test
+  public void shouldCreateCorrectFunctionCallExpressionWithSubscript() {
+    final String simpleQuery = "SELECT arraycol[0].name as n0, mapcol['key'].name as n1 FROM nested_stream;";
+    final Statement statement = KSQL_PARSER.buildAst(simpleQuery, metaStore).get(0);
+
+    final QuerySpecification querySpecification = getQuerySpecification(statement);
+    assertThat(querySpecification.getSelect().getSelectItems().size(), equalTo(2));
+    final Expression col0 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(0))
+        .getExpression();
+    final Expression col1 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(1))
+        .getExpression();
+
+    assertThat(col0, instanceOf(FunctionCall.class));
+    assertThat(col1, instanceOf(FunctionCall.class));
+
+    assertThat(col0.toString(),
+        equalTo("FETCH_FIELD_FROM_STRUCT(NESTED_STREAM.ARRAYCOL[0], 'NAME')"));
+    assertThat(col1.toString(),
+        equalTo("FETCH_FIELD_FROM_STRUCT(NESTED_STREAM.MAPCOL['key'], 'NAME')"));
+  }
+
+
+  @Test
+  public void shouldCreateCorrectFunctionCallExpressionWithSubscriptWithExpressionIndex() {
+    final String simpleQuery = "SELECT arraycol[CAST (item.id AS INTEGER)].name as n0, mapcol['key'].name as n1 FROM nested_stream;";
+    final Statement statement = KSQL_PARSER.buildAst(simpleQuery, metaStore).get(0);
+
+    final QuerySpecification querySpecification = getQuerySpecification(statement);
+    assertThat(querySpecification.getSelect().getSelectItems().size(), equalTo(2));
+    final Expression col0 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(0))
+        .getExpression();
+    final Expression col1 = ((SingleColumn) querySpecification.getSelect().getSelectItems().get(1))
+        .getExpression();
+
+    assertThat(col0, instanceOf(FunctionCall.class));
+    assertThat(col1, instanceOf(FunctionCall.class));
+
+    assertThat(col0.toString(), equalTo(
+        "FETCH_FIELD_FROM_STRUCT(NESTED_STREAM.ARRAYCOL[CAST(FETCH_FIELD_FROM_STRUCT(NESTED_STREAM.ITEM, 'ID') AS INTEGER)], 'NAME')"));
+    assertThat(col1.toString(),
+        equalTo("FETCH_FIELD_FROM_STRUCT(NESTED_STREAM.MAPCOL['key'], 'NAME')"));
+  }
+
+  private QuerySpecification getQuerySpecification(Statement statement) {
+    assertThat(statement, instanceOf(Query.class));
+    final Query query = (Query) statement;
+    assertThat(query.getQueryBody(), instanceOf(QuerySpecification.class));
+    return (QuerySpecification) query.getQueryBody();
+  }
+
+
+  @Test
+  public void shouldEnsureRewriteRequirementCorrectly() {
```

Comment:
good separation of tests!

(Not sure the "Incorrect rewrite requirement enforcement"  adds much)

Reply:
Made the message more specific.